src/OpenTelemetry.Api/Trace/TelemetrySpanNew.cs

```diff
@@ -0,0 +1,187 @@
+            this.activity?.AddTag(key, value);
+        }
+
+        /// <summary>
+        /// Adds a single <see cref="EventNew"/> to the <see cref="TelemetrySpanNew"/>.
+        /// </summary>
+        /// <param name="name">Name of the <see cref="Event"/>.</param>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void AddEvent(string name)
+        {
+            this.activity?.AddEvent(new EventNew(name).ActivityEvent);
+        }
+
+        /// <summary>
+        /// Adds a single <see cref="EventNew"/> to the <see cref="TelemetrySpanNew"/>.
+        /// </summary>
+        /// <param name="name">Name of the <see cref="Event"/>.</param>
+        /// <param name="timestamp">Timestamp of the <see cref="Event"/>.</param>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void AddEvent(string name, DateTimeOffset timestamp)
+        {
+            this.activity?.AddEvent(new EventNew(name, timestamp).ActivityEvent);
+        }
+
+        /// <summary>
+        /// Adds a single <see cref="EventNew"/> to the <see cref="TelemetrySpanNew"/>.
+        /// </summary>
+        /// <param name="name">Name of the <see cref="Event"/>.</param>
+        /// <param name="attributes">Attributes for the <see cref="Event"/>.</param>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void AddEvent(string name, IDictionary<string, object> attributes)
+        {
+            this.activity?.AddEvent(new EventNew(name, attributes).ActivityEvent);
+        }
+
+        /// <summary>
+        /// Adds a single <see cref="EventNew"/> to the <see cref="TelemetrySpanNew"/>.
+        /// </summary>
+        /// <param name="name">Name of the <see cref="EventNew"/>.</param>
+        /// <param name="timestamp">Timestamp of the <see cref="EventNew"/>.</param>
+        /// <param name="attributes">Attributes for the <see cref="EventNew"/>.</param>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void AddEvent(string name, DateTimeOffset timestamp, IDictionary<string, object> attributes)
+        {
+            this.activity?.AddEvent(new EventNew(name, timestamp, attributes).ActivityEvent);
+        }
+
+        /// <summary>
+        /// End the span.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void End()
+        {
+            this.activity?.Stop();
+        }
+
+        /// <summary>
+        /// End the span.
+        /// </summary>
+        /// <param name="endTimestamp">End timestamp.</param>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void End(DateTimeOffset endTimestamp)
+        {
+            this.activity?.SetEndTime(endTimestamp.UtcDateTime);
+            this.activity?.Stop();
+        }
+
+        /// <summary>
+        /// Makes the span as current.
+        /// </summary>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void Activate()
```

Comment:
This feels slightly out of place. I see `WithSpan` calls it, could we make it `internal`?

Reply:
You are right.